mlflow/transformers/signature.py

```diff
@@ -0,0 +1,193 @@
+from mlflow.utils.process import _IS_UNIX
+from mlflow.utils.timeout import MLflowTimeoutError, run_with_timeout
+
+_logger = logging.getLogger(__name__)
+
+
+_TEXT2TEXT_SIGNATURE = ModelSignature(
+    inputs=Schema([ColSpec("string")]),
+    outputs=Schema([ColSpec("string")]),
+)
+_CLASSIFICATION_SIGNATURE = ModelSignature(
+    inputs=Schema([ColSpec("string")]),
+    outputs=Schema([ColSpec("string", name="label"), ColSpec("double", name="score")]),
+)
+
+
+# Order is important here, the first matching pipeline type will be used
+_DEFAULT_SIGNATURE_FOR_PIPELINES = {
+    transformers.TokenClassificationPipeline: _TEXT2TEXT_SIGNATURE,
+    transformers.ConversationalPipeline: _TEXT2TEXT_SIGNATURE,
+    transformers.TranslationPipeline: _TEXT2TEXT_SIGNATURE,
+    transformers.FillMaskPipeline: _TEXT2TEXT_SIGNATURE,
+    transformers.TextGenerationPipeline: _TEXT2TEXT_SIGNATURE,
+    transformers.Text2TextGenerationPipeline: _TEXT2TEXT_SIGNATURE,
+    transformers.TextClassificationPipeline: _CLASSIFICATION_SIGNATURE,
+    transformers.ImageClassificationPipeline: _CLASSIFICATION_SIGNATURE,
+    transformers.ZeroShotClassificationPipeline: ModelSignature(
+        inputs=Schema(
+            [
+                ColSpec(DataType.string, name="sequences"),
+                ColSpec(DataType.string, name="candidate_labels"),
+                ColSpec(DataType.string, name="hypothesis_template"),
+            ]
+        ),
+        outputs=Schema(
+            [
+                ColSpec(DataType.string, name="sequence"),
+                ColSpec(DataType.string, name="labels"),
+                ColSpec(DataType.double, name="scores"),
+            ]
+        ),
+    ),
+    transformers.AutomaticSpeechRecognitionPipeline: ModelSignature(
+        inputs=Schema([ColSpec(DataType.binary)]),
+        outputs=Schema([ColSpec(DataType.string)]),
+    ),
+    transformers.AudioClassificationPipeline: ModelSignature(
+        inputs=Schema([ColSpec(DataType.binary)]),
+        outputs=Schema(
+            [ColSpec(DataType.double, name="score"), ColSpec(DataType.string, name="label")]
+        ),
+    ),
+    transformers.TableQuestionAnsweringPipeline: ModelSignature(
+        inputs=Schema(
+            [ColSpec(DataType.string, name="query"), ColSpec(DataType.string, name="table")]
+        ),
+        outputs=Schema([ColSpec(DataType.string)]),
+    ),
+    transformers.QuestionAnsweringPipeline: ModelSignature(
+        inputs=Schema(
+            [ColSpec(DataType.string, name="question"), ColSpec(DataType.string, name="context")]
+        ),
+        outputs=Schema([ColSpec(DataType.string)]),
+    ),
+    transformers.FeatureExtractionPipeline: ModelSignature(
+        inputs=Schema([ColSpec(DataType.string)]),
+        outputs=Schema([TensorSpec(np.dtype("float64"), [-1], "double")]),
+    ),
+}
+
+
+def infer_or_get_default_signature(
+    pipeline, example=None, model_config=None, flavor_config=None, timeout=60
```

Comment:
Should we make this timeout configurable? 1 minute seems relatively reasonable based on models that we use for CI, but some of the more complex and exceptionally large models that I've tested can take well over a minute to generate the final output. Having an environment variable that can be set to override the default might help to make this a bit more flexible.

Reply:
Yeah sounds good to me. I hardcoded it first for YAGNI, but no objection to add it if there is use case for it.